storage/innobase/buf/buf0dblwr.cc

```diff
@@ -602,20 +602,62 @@ static void buf_dblwr_check_block(const buf_page_t *bpage) noexcept
 }
 #endif /* UNIV_DEBUG */
 
+void buf_dblwr_t::print_info(FILE* file)
+{
+  mysql_mutex_assert_owner(&mutex);
+  slot *flush_slot= active_slot == &slots[0] ? &slots[1] : &slots[0];
+  std::ostringstream log_strm;
+
+  log_strm
+    << "--------------------\n"
+    << "Innodb: Double Write\n"
+    << "--------------------\n"
+    << "Batch running : " << std::boolalpha << batch_running << "\n"
+    << "Active Slot- "
+    << "first_free: " << active_slot->first_free
+    << " reserved: " << active_slot->reserved << "\n"
+    << "Flush Slot - "
+    << "first_free: " << flush_slot->first_free
+    << " reserved: " << flush_slot->reserved << "\n";
+  fputs(log_strm.str().c_str(), file);
+}
+
 bool buf_dblwr_t::flush_buffered_writes(const ulint size) noexcept
 {
   mysql_mutex_assert_owner(&mutex);
   ut_ad(size == block_size());
 
-  for (;;)
+  const int max_count= 60 * 60;
+  const int first_log_count= 30;
+  int log_count= first_log_count;
+
+  for (int count= 0;;)
   {
     if (!active_slot->first_free)
       return false;
     if (!batch_running)
       break;
-    my_cond_wait(&cond, &mutex.m_mutex);
-  }
 
+    timespec abstime;
+    set_timespec(abstime, 1);
+    my_cond_timedwait(&cond, &mutex.m_mutex, &abstime);
+
+    if (++count < first_log_count && !(count % 5))
+    {
+      ib::info() << "Long wait (" << count
+                 << " seconds) for double-write buffer flush.";
+      print_flush_info(stderr);
+      print_info(stderr);
+    }
+    else if (!(count % log_count))
+    {
+      ib::warn() << "Long wait (" << count
+                 << " seconds) for double-write buffer flush.";
```

Comment:
This needs to use `sql_print_warning()`. Shouldn’t this warning be included in `@global_suppression` in order to avoid spurious failures when some CI system is heavily loaded?